internal/pkg/agent/application/upgrade/details/details.go

```diff
@@ -0,0 +1,120 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package details
+
+import "sync"
+
+// Observer is a function that will be called with upgrade details
+type Observer func(details *Details)
+
+// Details consists of details regarding an ongoing upgrade.
+type Details struct {
+	TargetVersion string
+	State         State
+	ActionID      string
+	Metadata      DetailsMetadata
+
+	observers []Observer
+	mu        sync.Mutex
+}
+
+// DetailsMetadata consists of metadata relating to a specific upgrade state
+type DetailsMetadata struct {
+	ScheduledAt     string
```

Comment:
Should this be a `time.Time`?

Reply:
We don't plan to do any date/time math operations on this field so I made it a string, but I'll change it to `time.Time`.

Reply:
Done in 1f1940644dccaac10006890efbcafc63ddbe37ae.